Fuzzy matching

ta264 opened the pull request:
#### Database Migration
NO

#### Description
Add fuzzy matching to album and track names on import.

Fix matching albums when name is entirely special characters

#### Todos
- [x] Tests

#### Issues Fixed or Closed by this PR

This adds fuzz to album and track title matching when importing.  Means that the import is generally more successful.

This also fixes the case where the album name is entirely special characters, which means that `CleanName` is empty.

ta264 commented:
Hmm looks like Travis isn't managing to pull in the fuzzy string nuget package. Not sure why.

Qstick commented:
I’m a bit adverse to all the settings added as when users set to low it becomes a support nightmare and allows less advance users to alter parsing in such a way that it can effect their libraries adversely, do you think we could identify “optimal” values that would work for most cases

ta264 commented:
Yes, you're right about the config being dangerous. Currently you can go all the way from disabling fuzzy matching to "just pick the closest match even if its terrible" .

Currently the fuzz threshold setting determines how close the match must be in isolation. The gap setting determines how much better the best match must be vs the second best match to be accepted. 

I think the options we have are:
1) no config, just pick values we're happy with. 
2) binary option to enable/disable fuzz but with parameters we like. Might some users only want strict matching to avoid mistakes? 
3) allow config but only in a range that's sensible 
4) free config as we have now. Not a great idea. 

What would you guys prefer?

ta264 commented:
Talking about the config made me realise that the track matching could be improved.

At the moment the album matching compares the parsed title to all albums by the artist and can then choose the best. 

However the track matching looks at all tracks on the album with the correct track number and chooses the best one. However this means it's generally only looking at one possibility. I think it would be better to compare to all tracks on the album and then check that the best match has the correct track number. I think that would be less likely to give a wrong match.

Qstick commented:
I’m up for that as long as we are still using both. We got a lot of bad matches using one or the other separate as you could imagine.

ta264 commented:
I've got rid of the config and hard coded some values that I think are sensible.

I added some extra trace logging and in doing so realised that the nuget package I was using was calculating the final fuzzy match figure incorrectly so I've reworked it with a different package.

The track matches now works as suggested above.  It still insists that the track number has to be correct but makes sure that the name is also the best match if you ignore the track number, so it should be a bit less generous if there are a lot of similar track names.

Qstick commented:
Thanks @ta264 I'll test

Qstick commented:
Looks like this kills quite a few unit tests. We went from 16 failing to 46

ta264 commented:
Yeah should do. Didn't spot that. Will take a look.

On Sat, 6 Oct 2018, 17:23 Qstick, <notifications@github.com> wrote:

> *@Qstick* commented on this pull request.
>
> Will this work with our existing L Distance, instead of bringing in
> another library?
>
>
> https://github.com/lidarr/Lidarr/blob/cee1b33c60a1a9a17bce390fff3a0f2631f48d70/src/NzbDrone.Common/Extensions/LevenstheinExtensions.cs
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lidarr/Lidarr/pull/508#pullrequestreview-162262332>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEum73rqqDuBkbSqxhX6rMnlLb4xS0Swks5uiNkYgaJpZM4XGhOq>
> .
>

Qstick commented on src/NzbDrone.Core/Music/AlbumService.cs:
```diff
@@ -92,6 +93,11 @@ public Album FindByTitleInexact(string title)
             throw new NotImplementedException();
```
Can we purge this method now, since you have implemented below.

ta264 commented:
Done. I've also squashed it down to two commits and updated the description.

Qstick commented on src/NzbDrone.Core/Music/TrackService.cs:
```diff
@@ -100,6 +102,43 @@ select new
 
+        public Track FindTrackByTitleInexact(int artistId, int albumId, int mediumNumber, int trackNumber, string releaseTitle)
+        {
+            double fuzzThreshold = 0.6;
```
Why different threshold for Track than Album?

ta264 replied:
Simple answer is that was what's needed to get the test cases to pass.

I thought we could afford to be a bit less picky on the track name because we have an extra piece of information: the best track match has to have the right track number to be accepted but for albums we only have the name to go on.

For the examples that motivated this, the album name was generally pretty close (Eg it had CD appended) but track names can be a bit more off
